Feat: add list component

hollg opened the pull request:
Adds `List` component. Currently only supports unordered list with coloured bullets as that's what we need in the public Atom app. As soon as we have need/designs for an ordered list, it will be very easy to add with a `styled(ol, {/* ... */})` and a `type` prop for `List`.


<img width="110" alt="Screenshot 2021-03-10 at 08 53 33" src="https://user-images.githubusercontent.com/21319237/110602641-1c901880-817e-11eb-8001-032b58191e32.png">

hollg commented on src/components/list/List.tsx:
```diff
@@ -0,0 +1,48 @@
+        }
+      }
+    },
+    size: {
```
I added `size` as a variant because it `fontSize` affects the size of the bullets, so I thought it made sense to declare it once for the whole list. I'd be interested to hear any other thoughts on this though.

andoulla commented on src/components/list/List.test.tsx:
```diff
@@ -0,0 +1,19 @@
+    expect(container).toMatchSnapshot()
+    expect(await axe(container)).toHaveNoViolations()
+  })
+})
```
Not sure but should we have an empty list test too? (I have been caught out by empty states in the past

Also, we could check that it renders 2 `List.Items` (not sure again)

hollg replied:
If we test that two `List.Item`s are rendered here, wouldn't we just be testing React? We're adding those manually ourselves; there's no logic in the `List` to render them.

Re a test for an empty `List`: what would we test?

andoulla replied:
Maybe you are right, this way where it is just rending the children prob, there is no code in the component that handles the items. so yes maybe both my suggestions are redundant. I wasn't sure when I was asking it.

andoulla commented on src/components/list/List.tsx:
```diff
@@ -0,0 +1,48 @@
+})
+
+type ListProps = React.ComponentProps<typeof StyledUL> & {
+  theme?: 'primary' | 'secondary'
```
Quick q, but do we ever allow bullets that are the same colour as the text? Do we need to accommodate for that?

hollg replied:
No, not currently

andoulla replied:
Great, I was just checking we didn't miss anything

andoulla commented on src/components/list/List.mdx:
```diff
@@ -0,0 +1,14 @@
+title: List
+component: List
+description: List renders an unordered list
+---
```
Should the category be added here? content is a good fit for it

hollg replied:
Oh, yes, probably! I must have copied this from somewhere that was missing `category`. I'll add it now.

hollg commented on src/components/list/List.tsx:
```diff
@@ -0,0 +1,48 @@
+      primary: {
+        [`& ${StyledListItem}`]: {
+          '&::marker': { color: '$primary500' }
+        }
```
I think it seems like it should work, but I can't get it to ...

There's some unintuitive rule about using `&` when "chaining" (that's what one of the Modulz people called it when I asked for help on their Discord yesterday to get this to work at all). 

Using this seems to behave the same way that my original `&`-less attempt did, but I can't find a way to add the missing `&` in to your suggestion without it all styling disappearing.

hollg commented on src/components/list/List.tsx:
```diff
@@ -0,0 +1,48 @@
+  fontFamily: '$sans',
+  pl: '$2',
+  [`& ${StyledListItem}`]: {
+    pl: '$1',
```
Yeah, good question. I don't feel particularly strongly either way. I think there's a decent argument both for moving it (and thereby keeping the declaration for `StyledUL` smaller) and for keeping it here (and thereby keeping all the `::marker` shennanigans together, perhaps making it easier to grok what's going on).

Happy to move if you think makes more sense!

hollg commented:
@thomasdigby I've changed the `size` variants to match the styling of the `Text` components and updated the screenshot - could you take a look and let me know what you think?
